### PR TITLE
[PICARD-695] Tags are not properly cleared for some files even with the clear metadata option.

### PR DESCRIPTION
**Pull request by NCenerar**
Hello, I notice that in some case, the metadata were not competetly cleared even with the option enabled.
By saving the file without any metadata before saving with the metadata, it seems to work. I would like to provide the mp3 to illustrate the problem but the file is too big...

Thank you for your work, it is great ;)

### CODE REVIEW

**Comment by mineo** [comment]
Can you describe under which circumstances (settings, id3 version used for the old and for the new tags, ...) that makes this happen? At the moment this looks like an overly broad way solution to a possibly very specific problem.

**Comment by NCenerar** [comment]
Hello, I don"t know exactly the circustances that makes this to happen. In the tags options, I have enable "write tags to file", "clear existing tags", "remove id3 tags from FLAC files", "Remove APEv2tags from MP3 files", "Tag Compatibility 2.3", "ID3v2 Text Encoding UTF-16".

You can find the files that are causing the bug to appear at:
http://nicolas.cenerario.free.fr/bugged.zip
There are 6 files for 2 songs.
For each songs, the music data of the three corresponding files are the same (I made a byte-to-byte comparison after removing metadata from each files to verify). This mean that the only differences are on the metadata. But when I import the files into MusicBrainz Picard, the metadata are the same. I suspected the embeded pictures to be the cause of the differences. With the options I enabled, I expected that saving the files from MusicBrainz Picard would correct the metadata (including the embeded pictures also) but it didn't. In fact, I noticed that if I removed all metadata with MusicBrainz Picard and saving, it was working: the duplicated files were indeed the same, but if I left too many metadata, it was keeping something making each files different even if everything were supposed to be the same.

Since, the only "invisible" metadata were embeded pictures, I tried to remove them with another software (foobar 2000 which as a function to clear embded pictures), and indeed the files were identical as they should be. But maybe foobar 2000 is doing something else than just removing embeded pictures so, I'm not sure it is perfectly relyable.

Anyway, you were absolutely right. My solution was an overly broad way solution. I found a new very simple way to fix it. In fact, I really think it was a bug in MusicBrainz Picard because I saw that tags.clear() were used but in Mutagen API, it is tags.delete() that is documented. The file mutagen._util.py starts with:
"""Utility classes for Mutagen.

You should not rely on the interfaces here being stable. They are
intended for internal use in Mutagen only.
"""

Whoo, it is a long comment... Sorry. I hope I'm helping ;)

**Comment by NCenerar** [comment]
Oh, and a final comment:
I think the other files in picard.formats (apev2.py, asf.py, mp4.py, vorbis.py) should also be modified because they are using the same functions but I didn't do it because I don't have test files to run tests.
Should I updates them anyway or maybe insert a comment for future reference?

Thanks again!

**Comment by zas** [comment]
Can you please create (if needed, search existing issues first) a ticket in Picard bug tracker ? http://tickets.musicbrainz.org/browse/PICARD

It is better to describe the issue (and provide test files) in JIRA, than here, though this is the correct place to provide a patch.

**Comment by NCenerar** [comment]
Done : http://tickets.musicbrainz.org/browse/PICARD-695

**Comment by mineo** [comment]
The metadata of those files is not the same - they all declare different ID3 header sizes (People in the Shadow (Clear).mp3 2367 bits, People in the Shadow (Tagged 1).mp3 2775050 bits, People in the Shadow (Tagged 2).mp3 2776074 bits). Which software did you use to embed pictures in these files? I can't find an [APIC](http://id3.org/id3v2.3.0#Attached_picture) frame in these files and can't find any software that reads pictures from them.

**Comment by NCenerar** [comment]
Well, that is the problem. When I say that the files are the same, I mean that the audio data are the same and the displayed metadatas in Picard are the same.

As I mentionned, I suspected the enbeded pictures to be the source of the problems. And I am sorry but I really don't know which software put those invisible pictures. It might be iTunes, Windows Media Player or Sony Media Go... I don't know. I have been testing a lot of software to manage my music library and one of those must have done this...

Now, since I enable the option "clear existing tags", I am expecting the files to be identical after being saved from Picard. And that's where, in my point of view, there is a bug. The metadatas are not cleared. As you mentionned, it is probably because those invisible pictures are not found when deleting APIC frames. I tried that without any luck. I am no expert in audio file formats so, I'm sorry I cannot be more precise...

PS: I opened an issue for this, since the discussion is about the issue (and not about the patch), should we discuss there ?

**Comment by mineo** [comment]
Hm, so I've thought about this and am unable to form a clear opinion. On the one hand, calling `delete` is good because it's a standard API method of mutagen (although `clear` is [used internally](https://bitbucket.org/lazka/mutagen/src/113b8196614740b6f63b452d34d931eabdcce02c/mutagen/id3/__init__.py?at=default#cl-520) by `delete` in any case) and it cleans a bit more data (by relying on the ID3 header size declared by the file itself). On the other hand, this last part is really really dangerous: yours is the only case I've seen so far where there's a difference between what `clear` and `delete` remove from the file and that's because the ID3 header seems to include non-ID3 data. This could be audio data that we're deleting!

**Comment by NCenerar** [comment]
I understand your point of view but there is something very interesting and important to note:
After importing my files in Picard, if you delete all the tags within the list and save the files (so, the files are saved without any metadata because the "clear existing tags" option is enabled and there is no metadata field in the tagger), then, the files are properly cleared. The resulting files are the same as the ones obtained with my patch.
I don't know if my explanation is clear but that means that the "dirty" parts in my files are metadata parts and not audio ;)

In my point of view, there is no real danger in using delete, mainly because I believe that mutagen is good. It would be a critical bug of mutagen to publish a wrong function in its standard API. Now, if we want to analyse mutagen internal implementation: relying on the declared header size is a not a problem. It would be a problem only if the declared header size is wrong. But in this case, it's the file that is corrupted.

**Comment by phw** [comment]
I already commented on [PICARD-695](http://tickets.musicbrainz.org/browse/PICARD-695), here some additions.

Both using `clear` and `delete` will have a quite similar outcome and both basically fulfill what "Clear existing tags" promises (@NCenerar: more on that below). The difference is that `clear` only clears the metadata in memory while `delete` actually removes the entire ID3 block from the file which Picard will just recreate when saving the tags. The result is quite similar, but `delete` has the side effect to also reduce the ID3 padding in cases where it was unnecessarily large.

From a pure technical point of view I see the following problems with this patch:
  1. `clear` works completely in memory, while `delete` does directly modify the file.
  2. Using `delete` will always cause the entire file to get rewritten, and that two times (once for removing the ID3 block and once again to add it again). Using `clear` will only cause a rewrite if the new metadata is larger than the old metadata + padding.

Given that Picard already doesn't have the greatest reputation on the speed side I think this is a downside. Both points also increase the likelihood of file corruption, including damage to the audio part of the file.

From a conceptional point I see the problem that everybody seems to have another interpretation of what "Clear existing tags" does or should do:

  1. The current implementation: Before writing the new tags all existing tags are removed (the metadata, not any tag blocks in the file) and only the new tags Picard knows about are written to the file. The idea I think was that you can get rid of any extra metadata present in the file and you only get what Picard offers. Note that this interpretation focuses on the data side, not on the technical representation of this data inside a specific file format.
  2. Some people expect "Clear existing tags" to clear all existing tags except images ([PICARD-257](http://tickets.musicbrainz.org/browse/PICARD-257))
  3. Some people expect "Clear existing tags" to remove only the tags Picard knows about (I don't really get this one, as those are the tags that get overwritten anyway)
  4. Some people expect "Clear existing tags" to reduce the ID3 padding (this request)

Point 4 to me looks like really a different issue. I see why somebody wants to do it in cases like this one (where files have a padding of several MB size). But I don't see how it is related to removing the existing tags. I for example have "Clear existing tags" enabled because I do want to keep existing tags in the files Picard doesn't otherwise care about. But I maybe would want to remove the extra padding in cases like this, but without loosing existing tags.

I see the following possibilities for this:
  1. Either add an option to remove existing padding when saving
  2. Do it automatically once the padding reaches a certain sensible threshold.

I don't think mutagen offers any explicit API to modify the padding at the moment, but an implementation could probably just use `delete` by first loading the existing metadata, calling delete and writing the metadata back to the file.

**Comment by phw** [comment]
> yours is the only case I've seen so far where there's a difference between what clear and delete remove from the file and that's because the ID3 header seems to include non-ID3 data. This could be audio data that we're deleting!

As far as I could see the only difference in the files is extra padding. The ID3 header is huge in those files, but it's all null bytes. The data itself (both tags and audio) are identical. The files are ok, nothing is broken, they are from a technical view absolutely fine. They are just quite a bit larger than they need to be, maybe because gthe once included large APIC frames that were removed.

**Comment by zas** [comment]
Imho we could just add an option with a clear description that toggles between `clear` and `delete`.
Slower method can be handy sometimes to reduce size of files after reduction of cover art images size.

**Comment by NCenerar** [comment]
Should I add a checkbox "Remove extra padding" next to "Clear existing tags" ?
By the way, I implemented a plugin to add a context menu action "Save and rewrite header". Where should I submit it?

**Comment by zas** [comment]
> Should I add a checkbox "Remove extra padding" next to "Clear existing tags" ?

Yes, and it should only be enabled when "Clear existing tags" is checked.

> By the way, I implemented a plugin to add a context menu action "Save and rewrite header". Where should I submit it?

New plugin PR can be done in https://github.com/musicbrainz/picard-plugins project

**Comment by NCenerar** [comment]
So, I added a "remove_extra_padding" checkbox and it is visible only if "clear existing tags" is enabled (I would have prefered to lock the state of the "remove_extra_padding" checkbox but it seems that it is not possible...)

There is one thing I didn't do, it is translation. I would have liked to do it but couldn't find how. For learning purpose, can you give me some clue ?

Thank you.

File: picard/formats/asf.py [hunk 1]
```diff
@@ -41,16 +41,16 @@ def unpack_image(data):
     (type, size) = struct.unpack_from("<bi", data)
     pos = 5
     mime = ""
-    while data[pos:pos+2] != "\x00\x00":
-        mime += data[pos:pos+2]
+    while data[pos:pos + 2] != "\x00\x00":
```
Comment (zas) [inline picard/formats/asf.py]: Unrelated change. Please exclude those from the PR.

File: picard/formats/asf.py [hunk 2]
```diff
@@ -186,8 +186,11 @@ def _save(self, filename, metadata):
         log.debug("Saving file %r", filename)
         file = ASF(encode_filename(filename))
 
-        if config.setting['clear_existing_tags']:
-            file.tags.clear()
+        if config.setting["clear_existing_tags"]:
```
Comment (zas) [inline picard/formats/asf.py]: Instead it would be better to move those tests in a File method, and do something like:
```python
file.remove_tags()
```

Comment (NCenerar) [inline picard/formats/asf.py]: The procedure calls are not exactly the same so, I'm not sure it is relevant to try to do some factoring here. I think that it should be necessary to refactor all the File subclasses to have something more consistant.

Comment (phw) [inline picard/formats/asf.py]: From what I can see in this PR the calls in the various formats are always the same and that could go into the `file.remove_tags()` method:

```python
if config.setting["clear_existing_tags"]:
    if config.setting["remove_extra_padding"]:
        tags.delete()
    else:
        tags.clear()
```

Comment (NCenerar) [inline picard/formats/asf.py]: For apev2 and id3, yes, we have the same:
```python
if config.setting["clear_existing_tags"]:
    if config.setting["remove_extra_padding"]:
        tags.delete()
    else:
        tags.clear()
```
but for asf, mp4 and vorbis, we have
```python
if config.setting["clear_existing_tags"]:
    if config.setting["remove_extra_padding"]:
        file.delete()
    else:
        file.tags.clear()
```

Reading mutagen source code, it seems possible to change the `file.delete()` call into `file.tags.delete()` but I don't recommend it because it would skip a test.
Even in this case, there still would be a difference between `tags` and `file.tags`.
The best I can see would be a function in File like this:
```python
    def remove_tags(self, clear=lambda : None, delete=lambda : None):
        if config.setting["clear_existing_tags"]:
            if config.setting["remove_extra_padding"]:
                delete()
            else:
                clear()
```
and then, call it accordingly `self.remove_tags(tags.clear, tags.delete)` and `self.remove_tags(file.tags.clear, file.delete)`

But I'm not sure it would be really better. I can do it easily, I let you decide ;)

**Comment by NCenerar** [comment]
For the name, for now, I let "Remove extra padding" for the option and "remove_extra_padding" for the variable because I'm not sure what should be the link between the two.
I think that it is better to name the option according to what happens and not what is done because an option "Delete existing tags" next to "Clear existing tags" would be just confusing. Or an option like "Delete instead of clearing existing tags" would be just as confusing. Since there are already too many interpretations about "Clear existing tags", I believe that the option should be explicit. Since it is removing extra padding, I think it is better to keep that as a name.
Now, for the variable, I think that the same principle should apply. Because in fact, we don't really care what function is called, the important thing is that when the variable is set to true, the developper has to implement what is needed to remove the extra padding.

I think it is good practice to have "feature driven names" but if you want something else, just let me know ;)

**Comment by zas** [comment]
Please provide tests for this feature.

File: picard/formats/asf.py (hunk 2 above)
Comment (zas) [inline picard/formats/asf.py]: Hmmm, what about going the pythonic way, try a method, and if it fails try another one ?

Comment (phw) [inline picard/formats/asf.py]: > Hmmm, what about going the pythonic way, try a method, and if it fails try another one ?

I am not sure if this is good in this case. The method might exist, but have different behavior leading to hard to find bugs. Actually if it is format specific I would leave it inside the formats for best control over the behavior.

**Comment by Sophist-UK** [comment]
Sorry to chip in at a late stage, but I looked into this for another PR and I believe that the whole keep existing metadata thing is quite complicated, both conceptually and for implementation - and it seems to me that fixing this comprehensively might be easier in the long run than trying to sort out a specific issue.

So here are my thoughts:

# Requirements
1. Preserve existing tags where Picard has got replacement data from MB.
2. Preserve existing tags where Picard might get replacement data from MB, but in this instance it doesn't exist.
3. Preserve existing tags which Picard doesn't get data for ever (e.g. lyrics).
4. Delete tags where Picard has got replacement data from MB.
5. Delete tags where Picard might get replacement data from MB, but in this instance it doesn't exist.
6. Delete existing tags which Picard doesn't get data for ever (e.g. lyrics).

7. Users might want to pick and mix between the above.
8. For usability, we probably want a simple mode (keep or delete existing tags which Picard hasn't got replacement data for) and an advanced mode which provides the above.

9. Modify tags (i.e. do not reclaim space for tags which have changed - which will be quicker to write but use more space) or Rewrite file to reclaim the wasted space (slower but better results). Note: Not sure what capabilities Mutagen has in this area - so could be talking rubbish here.

# Implementation
If I remember correctly, the issue I was looking at (lyrics being deleted) was due to the keep / delete working on whether the new metadata object was based on the original metadata object or not - but the original metadata object only had the metadata objects that Picard could write not every existing tag. When you save the files, the code deletes all existing tags and then writes what is in the new metadata object - so anything that Picard isn't interested in is deleted.

So it seems to me that when we read a file, we need to put all tags into existing metadata object, and when we write the file we should clear or delete all tags and use what is in the new metadata object.

Does any of this make sense?

See PR #248 for the work I did which I haven;t had time to finish.

**Comment by NCenerar** [comment]
@zas I would like to provide tests but I'm no expert in audio files and I will need to investigate some more to create test data.

**Comment by NCenerar** [comment]
@Sophist-UK Hello and welcome in this complicated PR.
@phw had described the different interpretations for the "Clear existing tags" options and it clearly shows that Picard is missing something here.
I think that it would be nice to re-think this feature conceptually but it should be implemented step by step because it will probably involve a lot of modifications and I'm not sure a single PR is a good idea.

**Comment by NCenerar** [comment]
Hello, before implementing unit tests, I needed to create test files.
In the end, it gave me pretty interesting information that I'll try to share with you through a table.

But first, here is the thing I have done so that you can reproduce it:

1. For the options:
Check "Tags > **Clear existing tags**" and uncheck "Tags > **Remove extra padding**". Check "Cover Art > **Embed cover images into tags**".
2. Copy `./test/data/*.[flac,m4a,mp3,ogg,tta,wma,wv,mpc]` into `./test/picture1/`
3. Copy `./test/data/*.[flac,m4a,mp3,ogg,tta,wma,wv,mpc]` into `./test/picture2/`
4. Drag and drop `./test/picture1` and `./test/picture2` into a Picard instance (the audio files should be in the `Unmatched Files`)
5. Edit all tags from there so that every file has the same tags: I only set **Title: Title of the song** and **artist: The artist**.
6. Now, for the file in `./test/picture1/`, embed the picture of your choice by selecting each file individually, drag and drop the picture file you want from your explorer to the little cover art in Picard lower pane and save the file.
7. Repeat with another picture for the file in `./test/picture2/`
8. Duplicate folder `./test/picture1/` and `./test/picture2/` to get two new folders `./test/padding1/` `./test/padding2/` with the same content than the orginal.
9. In Picard, uncheck "Cover Art > **Embed cover images into tags**".
10. Drag and drop `./test/padding1` and `./test/padding2` into a Picard instance (the audio files should be in the `Unmatched Files`)
11. Save all the files.

Now, you should have for each file format, 4 files, 1 in each folder:

- `./test/picture1/`: Files with Picture 1 embedded
- `./test/picture2/`: Files with Picture 2 embedded
- `./test/padding1/`: Files with Picture 1 embedded then removed
- `./test/padding2/`: Files with Picture 2 embedded then removed

Now, here are the sizes of each file:

|Filename                    |File size|Is there a duplicate?| 
|----------------------------|--------:|---------------------|
|./test/picture1/test.flac   |29356    |No
|./test/padding1/test.flac   |29356    |No
|./test/padding2/test.flac   |14562    |No
|./test/picture2/test.flac   |14562    |No
|
|./test/picture1/test.m4a    |25054    |No
|./test/padding1/test.m4a    |25054    |No
|./test/padding2/test.m4a    |9694     |No
|./test/picture2/test.m4a    |9694     |No
|
|./test/picture1/test.mp3    |25044    |No
|./test/padding1/test.mp3    |25044    |No
|./test/padding2/test.mp3    |9684     |No
|./test/picture2/test.mp3    |9684     |No
|
|./test/picture1/test.ogg    |34643    |No
|./test/padding1/test.ogg    |5296     |Yes (1)
|./test/padding2/test.ogg    |5296     |Yes (1)
|./test/picture2/test.ogg    |14707    |No
|
|./test/picture1/test.tta    |24838    |No
|./test/padding1/test.tta    |24838    |No
|./test/padding2/test.tta    |9478     |No
|./test/picture2/test.tta    |9478     |No
|
|./test/picture1/test.wma    |29624    |No
|./test/padding1/test.wma    |7828     |Yes (2)
|./test/padding2/test.wma    |7828     |Yes (2)
|./test/picture2/test.wma    |14830    |No
|
|./test/picture1/test.wv     |24387    |No
|./test/padding1/test.wv     |2598     |Yes (3)
|./test/padding2/test.wv     |2598     |Yes (3)
|./test/picture2/test.wv     |9593     |No
|
|./test/picture1/test-sv7.mpc|23413    |No
|./test/padding1/test-sv7.mpc|1639     |Yes (4)
|./test/padding2/test-sv7.mpc|1639     |Yes (4)
|./test/picture2/test-sv7.mpc|8619     |No
|
|./test/picture1/test-sv8.mpc|23478    |No
|./test/padding1/test-sv8.mpc|1704     |Yes (5)
|./test/padding2/test-sv8.mpc|1704     |Yes (5)
|./test/picture2/test-sv8.mpc|8684     |No

My interpretation is that in the actual version of Picard, removing a cover replace it with dummy data (padding?) for flac, m4a, mp3 and tta files but not for the others format.

Do you want me to commit those data files? If yes, would it be a good place under `./test/data/embedded_pic/`?

**Comment by phw** [comment]
That's actually not that surprising, not all tagging formats make use of padding or at least mutagen does not support it for all formats. I think Ape tags don't support padding at all, but I could be wrong. For ASF there is [this mutagen ticket](https://bitbucket.org/lazka/mutagen/issue/201/asf-use-padding-object) for padding support. AFAIK Vorbis should support padding in theory, but there is [PICARD-32](http://tickets.musicbrainz.org/browse/PICARD-32) complaining that it is not always working. Those tickets are also interesting as they emphasize the importance of padding for performance reasons. That's especially true when writing lossless formats (due to the file sizes). That's the main reason for my initial objection to the patch as padding is indeed a feature and not a bug (though having the option to remove it is surely a useful feature under some circumstances).

> Do you want me to commit those data files?

I wouldn't commit those data files you created. Instead the unit tests should add large headers (e.g. cover art) to copies of the existing test files and then write smaller data again with and without your new option activated. The test should be verified by the expected change of file size. Have a look at the existing cover art tests at https://github.com/musicbrainz/picard/blob/master/test/test_formats.py#L786 ,  you should be able to adapt those.

**Comment by phw** [comment]
I did some digging and found some more interesting old tickets:

* [Ogg doesn't support padding](http://bugs.musicbrainz.org/ticket/5863) (contrary to FLAC) and people complained about slow writes.
* We already had the same request as this one for MP3 but for [FLAC files](http://bugs.musicbrainz.org/ticket/4125), complaining about the added padding. At the time it was closed as "won't fix".

**Comment by NCenerar** [comment]
I started to write some unit tests and I would like to have your comments before continuing.
For now, there is a lot of failure and I don't know if they should be considered "Expected". There is also a lot of Error which should be taken care of.

**Comment by NCenerar** [comment]
I updated the unit tests and here is my conclusion:

- AIFF and MP3 files: All is ok.
	- Can embed GIF, JPG and PNG.
	- Embedding an picture and clearing it effectively creates a padding as expected.
	- Using `delete` remove the padding.

- FLAC files: Cannot remove padding!
	- Can embed GIF, JPG and PNG.
	- Embedding an picture and clearing it effectively creates a padding as expected.
	- Using `delete` DO NOT remove the padding.

- M4A files: Cannot embed GIF. Cannot remove padding!
	- Can embed JPG and PNG but not GIF.
	- Embedding an picture and clearing it effectively creates a padding as expected.
	- Using `delete` DO NOT remove the padding.

- OGG, TTA, WMA, WV, SV7.MPC, SV8.MPC files: Cannot create padding!
	- Can embed GIF, JPG and PNG.
	- Embedding an picture and clearing it DO NOT create a padding as expected!

**Comment by phw** [comment]
Nice results.
  * AIFF, MP3: Everything fine and as we expected
  * FLAC: I think this is due to the fact that padding is in this case a feature of the FLAC format, not the tagging format. No idea if we can do something about this.
  * M4A: I have no idea how padding is supposed to work here
  * OGG, WMA, APE: See my earlier comments, but the results are as expected as padding is not supported.

What we can say is that using the `delete` method is no reliable way to remove the padding and works only for ID3. Not sure how we should proceed in this case, but the current implementation of having "remove_extra_padding" as a suboption to "clear_existing_tags" seems wrong if we don't get it to work with other formats supporting padding. One option would be to have this option explicitly only for ID3 ("Remove padding when saving ID3 tags").

Personally I am not really in favor of having this option at all given the current technical state. I'd prefer a clean, non destructive way of removing just the padding (independent of "clear_existing_tags") instead of the hackish `delete`.

File: test/test_formats.py [hunk 3]
```diff
@@ -1062,3 +1065,451 @@ def _test_cover_art_types_only_front(self, filename, expect):
...
+                self.assertPicNumberEqual(metadata_4, nb_pics_3)
+                self.assertPicNumberEqual(metadata_4, nb_pics_2 - 1)
+                self.assertFileSizeLess(aud, size_3)
+
+    def assertRemoveExtraPaddingDoesNotThrowException(self, f):
+        with self.tempcopy(f) as ff:
+            config.setting.update({'clear_existing_tags': True})
+            config.setting.update({'remove_extra_padding': True})
+            pff = picard.formats.open(ff)
+            metadata = pff._load(ff)
```
Comment (zas) [inline test/test_formats.py]: AttributeError: 'NoneType' object has no attribute '_load' repeated when running tests.

**Comment by zas** [comment]
According to the tests, and the discussion, it seems we will not apply this PR.

Can this feature be implemented as a post-save hook by a plugin calling external padding removal tools ? If Mutagen proposes a reliable method to do so we could easily modify the plugin, and eventually merge it back to main code if needed.

**Comment by phw** [comment]
I have filed a mutagen ticket: https://bitbucket.org/lazka/mutagen/issues/229/add-a-way-to-reduce-the-padding-after

**Comment by phw** [comment]
As of mutagen 1.31 there is now padding support for ASF and OGG. Also Mutagen will reduce the padding on save if a certain threshold is reached (by deafault padding is removed if it takes up more than 2.5% of the entire file). The API also allows more control over that behavior.

My suggestion would be to close this PR and rely on the Mutagen default. If we later find out the defaults don't work well for some people we can improve on this. But at least we have now a clean API to work with.
